kotlinx-coroutines-core/common/src/Delay.kt

```diff
@@ -154,8 +154,15 @@ public suspend fun delay(duration: Duration): Unit = delay(duration.toDelayMilli
 internal val CoroutineContext.delay: Delay get() = get(ContinuationInterceptor) as? Delay ?: DefaultDelay
 
 /**
- * Convert this duration to its millisecond value.
- * Positive durations are coerced at least `1`.
+ * Convert this duration to its millisecond value. Durations which have a nanosecond component less than
+ * a single millisecond will be rounded up to the next largest millisecond.
  */
-internal fun Duration.toDelayMillis(): Long =
-    if (this > Duration.ZERO) inWholeMilliseconds.coerceAtLeast(1) else 0
+internal fun Duration.toDelayMillis(): Long {
```

Comment:
it will fail on attempt to do `Duration.INFINITE.toDelayMillis()` as the `Duration.INFINITE.inWholeMilliseconds` returns Long.MAX_VALUE. Adding 1 to it will overflow and return Long.MIN_VALUE. Another place when it may overflow is `millis * nanosecondsInMillisecond` (for any millis duration above Long.MAX_VALUE / 1000_000). 

Probably the easiest way to round it up is to defer all the math to kotlin.time
```
(this + 999999.nanoseconds).inWholeMilliseconds
```

Reply:
Switched to use modulo arithmetic.

Reply:
It will now fail for any large (but finite) number, just because of the precision error when attempting to access `inWholeNanoseconds`

i.e. this one should fail
```
@Test fun large_but_finite() {
    val milliseconds = Long.MAX_VALUE / 10
    assertEquals(
        expected = milliseconds, 
        actual = milliseconds.milliseconds.toDelayMillis(),
    )
}
```
because 
```
Long.MAX_VALUE                | 9223372036854775807
Long.MAX_VALUE/10             | 922337203685477580
In nanos (absolute precision) | 922337203685477580000000 | mod1000000=0
In nanos (coerced to 64bits)  | 9223372036854775807      | mod1000000=775807

```

Reply:
I see. Good catch. Will delegate to kotlin.time arithmetic.

Reply:
Updated.

Reply:
I agree with @denis-bezrukov:
> `(this + 999999.nanoseconds).inWholeMilliseconds`

seems to be the cleanest and, I suspect, also the most efficient solution available to us without getting access to the internals of `Duration`. Here, in realistic cases, it's just a single addition and a conversion to milliseconds.

Reply:
Pushed. Thank you two for the review.